### PR TITLE
Issue #15985: Expand Default Value of constantWaiverParentToken in MagicNumberCheck with All Operators

### LINKED ISSUE #15950: False positive: magicnumbers with bit shift expression
I have read check documentation: https://checkstyle.org/checks/coding/magicnumber.html
I have downloaded the latest cli from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
$ javac Test.java
$
```

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="severity" value="error"/>
  <property name="fileExtensions" value="java, properties, xml"/>
  <module name="TreeWalker">
    <module name="MagicNumber">
       <property name="ignoreNumbers" value="-2,-1,0,1,2,100"/>
       <property name="ignoreHashCodeMethod" value="true"/>
       <property name="ignoreFieldDeclaration" value="true"/>
       <property name="severity" value="warning"/>
       <property name="id" value="checkstyle:magicnumber"/>
    </module>
  </module>
</module>
```
```java
public class Test {
    public static final int BIT0 = 1;
    public static final int BIT1 = 1 << 1;
    public static final int BIT2 = 1 << 2;
    public static final int BIT3 = 1 << 3;
    public static final int BIT4 = 1 << 4;
}
```
```bash
$ java -jar checkstyle-10.20.1-all.jar -c config.xml Test.java 
Starting audit...
[WARN] /var/tmp/Test.java:5:41: '3' is a magic number. [checkstyle:magicnumber]
[WARN] /var/tmp/Test.java:6:41: '4' is a magic number. [checkstyle:magicnumber]
Audit done.
```

I would have expected bit-shift operators to also be excluded from this check, just like `2 * 2 * 2` would work without checkstyle complaining...

### LINKED ISSUE #15985: Expand Default Value of `constantWaiverParentToken` in MagicNumberCheck with All Operators
Check documentation: https://checkstyle.org/checks/coding/magicnumber.html#MagicNumber

**Rationale of the check:** 
> Checks that there are no ["magic numbers"](https://en.wikipedia.org/wiki/Magic_number_%28programming%29) where a magic number is a numeric literal that is not defined as a constant.

> constantWaiverParentToken: Specify tokens that are allowed in the AST path from the number literal to the enclosing constant definition.

The rationale of the check is to violate all magic numbers that are not defined as a constant and `constantWaiverParentToken`
specify tokens allowed in enclosing constant definition.

I suggest expanding the allow list with all operators (bitwise, comparison, arithmetic, logical, etc..) because:

- The property concerns only constant fields, which should be always ignored.
- To maintain consistency suggest ignoring all operators by default.
- This would align with the rationale that numbers inside constant fields are inherently non-magic numbers.

```Java
public class Example {
    public static final int CONSTANT = 7 * 6; // Consistency in ignoring all operators
    public static final boolean CONDITION = 7 >> 6; // Consistency in ignoring all operators
}
```

This simplifies check configuration for users and ensures that all constant fields are treated uniformly, regardless of the operators used.
Also, users have the option to override this property with any tokens that they need. with common sense, I would say that all operators should be included here.

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #15985

### CODE REVIEW

**Comment by mahfouz72** [comment]
Github, generate report for MagicNumber/Example1

**Comment by romani** [comment]
Github, generate report for MagicNumber/Example1

**Comment by romani** [comment]
@mahfouz72, please try to trigger report one more time, not clear why your execution failed.

**Comment by mahfouz72** [comment]
Github, generate report for MagicNumber/Example1

**Comment by romani** [comment]
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/2cad2ada63b12f7151ee812317c057227694c7e2_2024105158/reports/diff/Vavr/index.html#A1

```
@Test
public void shouldCheckHashCodeInLeafList() {
    HashArrayMappedTrie<Integer, Integer> trie = empty();
    trie = trie.put(0, 1).put(null, 2);  // LeafList.hash == 0
    final Option<Integer> none = trie.get(1 << 6);  // (key.hash & BUCKET_BITS) == 0
    assertThat(none).isEqualTo(Option.none());
}
```

I am not sure  that removal of this violation is good. but in the same time if there would be `*` instead of `<<`, there will be no violation, so making `<<` so special is not very good.

**Comment by mahfouz72** [comment]
> kind of magic number, but it is field, so should be ignored if ignoreFieldDeclaration=true, but this report is for default config, 

This is ignored because the field is constant, the same as we ignore `/` (BASE / 5, for example), so `|` has nothing special to be a violation.

I agree with extending the allow list to make the check less strict. Users can still customize the list if needed.

please let me know your final decision.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java:8]: We changed default, but we didn't change all other  Inputs with "(default)".
Please do.
I am not sure in what issue we will do validation of default, I will investigate later on.

**Review by romani — CHANGES_REQUESTED** [review]
Do you know what other operators are not this waiver?
I just try to formulate some reasons of this property and tokens presence in it.
To make it clear what is default and what users should do themselves.

Items:

**Comment by romani** [comment]
Can we suggest user in issue a workaround by setting of this property with required tokens?

**Comment by romani** [comment]
Can we suggest user in issue a workaround by setting of this property with required tokens?

**Comment by romani** [comment]
Can we suggest user in issue a workaround by setting of this property with required tokens?

**Comment by mahfouz72** [comment]
> Do you know what other operators are not this waiver?

There are a bunch of other operators that are not on the list such as mod operation, comparison operators, and conditional operators

Example:

```
D:\CS\test
cat src/Test.java
public class Test {

    public static final boolean BIT6 = 1 >= 3; // violation
    public static final boolean BIT7 = 1 != 3; // violation
    public static final int BIT8 = 1 % 3; // violation
    public static final int BIT9 = 1 > 2 ? 6 : 5; // 2 violations
}
D:\CS\test
java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:3:45: '3' is a magic number. [MagicNumber]
[ERROR] D:\CS\test\src\Test.java:4:45: '3' is a magic number. [MagicNumber]
[ERROR] D:\CS\test\src\Test.java:5:40: '3' is a magic number. [MagicNumber]
[ERROR] D:\CS\test\src\Test.java:6:44: '6' is a magic number. [MagicNumber]
[ERROR] D:\CS\test\src\Test.java:6:48: '5' is a magic number. [MagicNumber]
Audit done.
Checkstyle ends with 5 errors.
```

> I just try to formulate some reasons of this property and tokens presence in it.
To make it clear what is default and what users should do themselves.

I still can't formulate reasons why we add some operators and not others. I don't think of any use case when we could violate magic numbers with one operator but not others.

The thing here is consistency. If one operator triggers a magic number violation, why should others (like addition or subtraction) not? Can we say that some operators (like arithmetic ones) are more commonly used with numbers That is why we added them in default list but others are not commonly used so we leave the option for users to add them to the list.

**Comment by romani** [comment]
We can make any rule we think is reasonable. Let's drop few ideas/reasons and choose by vote what is more clear to active maintainers.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java:8]: https://github.com/checkstyle/checkstyle/issues/15983 is created to improve separately.

**Comment by mahfouz72** [comment]
> We can make any rule we think is reasonable. Let's drop few ideas/reasons and choose by vote what is more clear to active maintainers.

moved to https://github.com/checkstyle/checkstyle/issues/15985

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java]: @mahfouz72 , please update other Inputs with default on this property to have such tokens.

**Comment by romani** [comment]
@mahfouz72 , see https://github.com/checkstyle/checkstyle/pull/15957#discussion_r1874500779

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java
Comment (mahfouz72) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java]: This will be a huge update we have many inputs for this check :)

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java]: We need it, I hope it is just copy paste

Comment (mahfouz72) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java]: Done. It was painful because I had to change row numbers in every unit test

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java]: Hm, we always keep 2 empty lines in config comment to avoid line shift.

Comment (mahfouz72) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java]: The addition is 2 lines, and I can't remove the 2 empty lines in the config comment because there will be no empty lines at all and we have a test for this. 

It will be good if we had 3 empty lines not 2

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/magicnumber/InputMagicNumberWithBitwiseOperatorsWithIgnoreFieldsFalse.java]: Ok, next time try to reuse empty line in header comment, if post.

**Review by romani — APPROVED** [review]
Ok to merge

**Review by nrmancuso — APPROVED** [review]
Thanks @mahfouz72 !

**Pull request merged**
